### PR TITLE
Document branches and tags

### PR DESCRIPTION
**Pull request by icecrime**
Following up on #14445, this PR documents branches and tagging. No changes there, it's just about documenting what we do for other Docker projects to reuse.

Contributes to #14662.

### CODE REVIEW

File: project/BRANCHES-AND-TAGS.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+The `release` branch contains the last _released_ version of the code for the project.
+
+The `release` branch is only updated at each public release of the project. The mechanism for this
+is that the release is materialized by a pull request against the `release` branch which lives for
+the duration of the code freeze period. When this pull request is merged, the `release` branch gets
+updated, and its new state is tagged accordingly.
+
+# Tags
+
+Any public release of a compiled binary, with the logical exception of nightly builds, should have
```
Comment (stevvooe) [inline project/BRANCHES-AND-TAGS.md]: All tags should be signed by a maintainer with PGP.

File: project/BRANCHES-AND-TAGS.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+Docker projects should normally have only three living branches at all time, including the regular
+`master` branch:
+
+## `docs` branch
+
+The `docs` branch support documentation updates between product releases. This branch allow us to
+decouple documentation releases from product releases.
+
+## `release` branch
```
Comment (stevvooe) [inline project/BRANCHES-AND-TAGS.md]: Typically, we have a release branch per major version. For instance, we have `release/2.0` and we'll have `release/2.1`. This allows doing patch releases on individual branches.

Comment (stevvooe) [inline project/BRANCHES-AND-TAGS.md]: Note, we make the release tag on master, then take the release branch from that tag. This may not work for the main docker project. Changes that land in both branches should be merged to both. In the past, we didn't cherry-picking, but that was a huge pain for 2.0.1.

File: project/BRANCHES-AND-TAGS.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+Note: details of the release process for the Engine are documented in the
+[RELEASE-CHECKLIST](https://github.com/docker/docker/blob/master/project/RELEASE-CHECKLIST.md).
+
+# Branches
+
+Docker projects should normally have only three living branches at all time, including the regular
+`master` branch:
+
+## `docs` branch
```
Comment (stevvooe) [inline project/BRANCHES-AND-TAGS.md]: I know this is the current process, but is there any reason to have a separate docs branch from the release branch? We should publish docs from release/2.0, release/2.1, etc.

Comment (icecrime) [inline project/BRANCHES-AND-TAGS.md]: Because we want to be able to release docs in between product release (example: fix a broken link).

File: project/BRANCHES-AND-TAGS.md (hunk 1 above)
Comment (icecrime) [inline project/BRANCHES-AND-TAGS.md]: And blood. Don't forget about blood.

Joke aside, I'm not opposed. Any opinion @jfrazelle @tianon?

File: project/BRANCHES-AND-TAGS.md (hunk 2 above)
Comment (icecrime) [inline project/BRANCHES-AND-TAGS.md]: But what's the use of long running branches as long as you have the tag?

File: project/BRANCHES-AND-TAGS.md (hunk 3 above)
Comment (stevvooe) [inline project/BRANCHES-AND-TAGS.md]: That can be a minor fix to a release branch without making a tag.

File: project/BRANCHES-AND-TAGS.md (hunk 2 above)
Comment (stevvooe) [inline project/BRANCHES-AND-TAGS.md]: Generally, documentation fixes. Having a branch for each minor let's you do security releases going back several versions. They don't actually need to exist unless a point release is actually made, especially if the tag is made from master.

The main benefit is that you don't need to force push over the release branch after each release if the the release branch was made with a cherry pick or rebase. Each release becomes a fresh branch off master.

Comment (dmp42) [inline project/BRANCHES-AND-TAGS.md]: The main difference between engine and projects like notary or registry is the number of releases that are maintained in the long run. Engine de facto maintain only the last release, while others maintain at least the last two or three versions. This does change perspective a bit on branching.

Is it really important to harmonize things on this point?

File: project/BRANCHES-AND-TAGS.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+is that the release is materialized by a pull request against the `release` branch which lives for
+the duration of the code freeze period. When this pull request is merged, the `release` branch gets
+updated, and its new state is tagged accordingly.
+
+# Tags
+
+Any public release of a compiled binary, with the logical exception of nightly builds, should have
+a corresponding tag in the repository.
+
+The general format of a tag is `vX.Y.Z[-suffix[N]]`:
```
Comment (thaJeztah) [inline project/BRANCHES-AND-TAGS.md]: Should we use `major`, `minor`, `patch` here, instead of `X`, `Y`, `Z`?

File: project/BRANCHES-AND-TAGS.md (hunk 1 above)
Comment (thaJeztah) [inline project/BRANCHES-AND-TAGS.md]: sgtm

File: project/BRANCHES-AND-TAGS.md (hunk 3 above)
Comment (moxiegirl) [inline project/BRANCHES-AND-TAGS.md]: The `docs` branch supports..

File: project/BRANCHES-AND-TAGS.md (hunk 1 above)
Comment (diogomonica) [inline project/BRANCHES-AND-TAGS.md]: I think this is silly:
- No one validates GPG signatures unless it's automated
- We don't have a good method for securely distributing these keys or rotating them
- Keys based on individuals mean that they are a lot more prone to being lost, compromised, or the individual stopping being a maintainer, making the set of trusted keys a lot more dynamic and this whole thing even more useless.

If you want trust over every tag, have both @tianon and @jfrazelle tweet them after pushing. That will be a lot more secure. You can even call it twoeet-factor-authentication.

File: project/BRANCHES-AND-TAGS.md (hunk 4 above)
Comment (diogomonica) [inline project/BRANCHES-AND-TAGS.md]: Also document the set of possible N's

File: project/BRANCHES-AND-TAGS.md (hunk 2 above)
Comment (moxiegirl) [inline project/BRANCHES-AND-TAGS.md]: It is important to the documentation management process to keep the `docs` branch. We have to coordinate across multiple repositories and releases on tight schedules.  Having a different process in different repositories or tracking multiple/unique branches depending on the repository just adds a management burden to docs --- and we are already short staffed.

File: project/BRANCHES-AND-TAGS.md (hunk 1 above)
Comment (jfrazelle) [inline project/BRANCHES-AND-TAGS.md]: We do this anyways IIRC the "official announcement" of 1.7.0 was my tweet because their was no blog post.... But you are also not a big gpg fan and I am so I like the idea of signing the tags, the kernel does it

File: project/BRANCHES-AND-TAGS.md (hunk 2 above)
Comment (icecrime) [inline project/BRANCHES-AND-TAGS.md]: We don't only maintain the last release: we do security patches on version N and N-1.

File: project/BRANCHES-AND-TAGS.md (hunk 1 above)
Comment (jfrazelle) [inline project/BRANCHES-AND-TAGS.md]: i do agree with the fact that individual keys could be lost but this is also the way debian and a bunch of other project i think even tor sign releases or packages etc

**Comment by spf13** [comment]
So we have a few goals here and I feel like we're not being clear on them and the potential solutions being suggested.. I'm not sure I know them all but I'll list the ones I believe we have.

Goals:
* Standardize the organization across repos
* Decouple doc releases from software releases
* Keep the overhead/management of software releases at a minimum
* Keep the overhead/management of doc releases at a minimum
* Easily maintain docs for multiple releases (including updates post release)
* Make it clear which branch is the "current release" one.

From this discussion it seems like the solution @stevvooe is suggesting accomplishes most of this and in a pretty clean fashion. 

The only disadvantage I see with his solution is that scripting things that run on the latest release branch may be trickier than the other approach with a "release" and a "docs" branch with the fixed names.

Am I missing anything?

**Comment by moxiegirl** [comment]
My major concern is around the tooling.  I have been releasing interim docs at least once and often more frequently since Dockercon. I was doing it perhaps once every other week since 1.6. Our current release tooling requires me to enter a value for the branch to pull from. I literally type that value in. If the value is the same across all the repos, ie. `docs`, and is stable, then I have a no-touch task when it comes to releasing between versions.

If that value can change, meaning I need to include release tag 1.0X for machine and something else for compose...and so on. Moreover --- I have to check the repo, confirm with a human being I'm pulling the right tag blah, blah, blah then it is more prone to error as I have to do a lot of manual, high-touch steps.

Moreover a complex balance of varying version across multiple repos becomes much more cumbersome a process to manage, document, and teach.  If we intend to grow to the size the company seems to be --- I'd prefer to keep it simple even if the elegance of the solution is lacking.

Now,  @mikedougherty did the work coming up with the release tooling. As a release engineer familiar with our infrastructure and the docs in particular, he might be suited to cutting this knot.

**Comment by mikedougherty** [comment]
I don't know that I can necessarily end this discussion but maybe I can clarify what I see is the remaining issue, and how it affects future tooling.

@spf13 did a good job of laying out the desired effects. There definitely needs to be a long-running branch per supported release ('supported', in this context of documentation, might be a bit more loose -- consider the example of a typo fix for the engine 1.2 docs).

After that, there are two possibilities:
 * `release/1.7` branch with both docs changes and the actual code being released. In this model, patch releases would continue to be tagged on this branch and the HEAD of the branch may not match the latest patch release due to documentation changes. (This seems to be @stevvooe's position)
 * `release/1.7` and `release/1.7-docs` branches. Some folks prefer a more 'pure' approach to release branches -- only tagged commits should ever be the HEAD of a release branch. Ad-hoc docs changes would be released from the `releases/1.7-docs` branch. (This seems to be @icecrime's position)

Apologies to @icecrime and @stevvooe if I've misrepresented either of you :)

I have a slight preference for the first, but only due to past familiarity. In practice, any future automation/tooling could handle both scenarios easily.

It's also not a terrible idea to have a simple `release` (and possibly `docs`) branch which tracks the 'active' versioned release/docs branches. This has its it benefits for automation as @spf13 and @moxiegirl have mentioned. It can be a bit more work (and therefore more error-prone) on the dev side, but worth pointing out is that any errors made are likely to be low-impact and easily fixed.

**Comment by stevvooe** [comment]
@mikedougherty You've represented the suggestions accurately.

@moxiegirl Maintaining a separate `docs` and `release` branch for "automation" doesn't make a lot of sense. It trades off some automation in the docs process for a lot of manual work getting these branches in the right state. Part of the issue is rebasing and cherry picking into `release` with each release. Effectively, the branch head is always changing. With per release, branches we avoid these issues. By using standardized [semantic versioning](http://semver.org/) for the major and minor on the branch name,  you have an ordering defined over release branches that doesn't need to be maintain manually.

And to clarify, I am not advocating "full purity" in release branches. This doesn't buy us much. We just want the docs in the HEAD of the each "active" release branch to be releasable. It's okay if it contains a bunch of doc fixes on top of the tag. If another release comes out, we just tag that HEAD and move along. Docs are always up to date and we have them per release, going all the way back. If we have to backport doc fixes across all release branches, we just do it to each branch and job done.

**Comment by moxiegirl** [comment]
I'd prefer to keep everything as it is for the 1.8 release. We can make changes after if necessary.

File: project/BRANCHES-AND-TAGS.md (hunk 1 above)
Comment (tianon) [inline project/BRANCHES-AND-TAGS.md]: Yeah, I tried to get the maintainers to get PGP keys set up a good long while back so we could start signing releases this way (because verifying binaries via automated processes is a huge +1 in my book -- there are several places I use the raw binaries that I'd love to be able to verify them), but the take rate was low.

All that aside, I think this would be much easier to manage if we had an official "release server" though (like we've been discussing) so that these keys don't have to be shared around more than the official releasing server, and so people don't have to import every single maintainer's personal keys to verify the binaries (or find out which maintainer was the "release maintainer" for a particular release).  Python signs their artifacts with the personal key of the release maintainer and it's a real PITA to try and find which key signed which release.

TL;DR, I'm not opposed, but I don't know that just having the release manager use their own personal key is really tenable, and sharing keys isn't really great either.

**Comment by thaJeztah** [comment]
Been following this discussion and, although I see benefits in having the docs as part of the release branch, the intent of this PR was to document what we do *currently*.

I suggest to focus on that in this PR, and open a new issue for improving / discussing changes to the branching separately.

**Comment by stevvooe** [comment]
@thaJeztah Agreed but it also has the implication that other projects _must_ adopt this approach. The absurdity here is that we a forced to do a lot more work to maintain `docs` and `release` to support the documentation workflow.

We spent a bunch of time cherry picking and rebasing into docs and release for the last release. It was error prone and we will not be following that process in the next release.

**Comment by thaJeztah** [comment]
@stevvooe I don't think this document *enforces* other projects to follow this approach ("it's just about documenting what we do for other Docker projects to reuse"). 

(However, it *would* be good to reach consensus on a uniform approach across the repos)

I'm **not** against changing the current model, but also practical; making changes in the current model will (in practice) come down to @moxiegirl. Given that she indicated that she would like to keep the current model for the 1.8 release, I like to respect her opinion here.

For that reason, I'd  like to merge this (after addressing comments) and move the discussion for the larger changes elsewhere. Hopefully we can reach consensus soon and make all the preparations / tooling changes so that we can start using the (new?) branching model directly after 1.8 has been released.

File: project/BRANCHES-AND-TAGS.md (hunk 1 above)
Comment (jfrazelle) [inline project/BRANCHES-AND-TAGS.md]: let's skip it

**Comment by jfrazelle** [comment]
lets end the bike shed and mergeeeeeee

**Comment by thaJeztah** [comment]
@jfrazelle agreed; wdyt: should we have https://github.com/docker/docker/pull/14720/files#r34953304 and https://github.com/docker/docker/pull/14720#discussion_r34959617 addressed in this PR, or a follow up?

**Comment by jfrazelle** [comment]
thsi PR is probably fine

**Comment by thaJeztah** [comment]
ping @icecrime could you address https://github.com/docker/docker/pull/14720#discussion_r34959617, then I think this can be merged

File: project/BRANCHES-AND-TAGS.md (hunk 1 above)
Comment (stevvooe) [inline project/BRANCHES-AND-TAGS.md]: How can this possibly be controversial? All you do is add a flag to the `git tag` command. It takes two seconds to generate a PGP key and another few to back it up.

If it is useless, we've lost nothing.

File: project/BRANCHES-AND-TAGS.md (hunk 3 above)
Comment (SvenDowideit) [inline project/BRANCHES-AND-TAGS.md]: I would also rather the docs be published from the already existing release branch - having them separate generally means the code in that branch does not get updated when point releases happen (and we don't do docs point release branches) and by having them separate, both the docs and release branches are not fully up todate.

However - this PR is about documenting what we do today :) so this change would go elsewhere.

**Comment by icecrime** [comment]
Fixed a typo, clarifies that this is about Docker Engine. You entirely lost me with the rest of the conversation: I'm just documenting what we do.

If it correctly depicts what we do today, please merge. If you want to change the process, please merge and submit a PR against this file.

**Comment by thaJeztah** [comment]
@icecrime yeah, it wasn't very important, just a request to be more explicit what X, Y and Z were, and what suffixes to use (alpha/beta/rc). No biggie - can be left for a follow-up.

Merging

**Comment by thaJeztah** [comment]
Oh, and LGTM :smile:

For the follow-ups:

- @stevvooe @diogomonica @tianon @jfrazelle could one of you create a separate issue / PR where we can discuss the signing of releases?
- @mikedougherty @stevvooe @moxiegirl could one of you create a separate issue / PR for a new branching model? I think there's interest in this, but we need to find an approach that suits us all, a migration plan, and (probably) changes around tooling and procedure :+1:
